### ISSUE TITLE
Add cascading dashboard filters for joined fields

### ISSUE THREAD

**Issue by jpetey75** [lightdash/lightdash#10911]
### Description

We need to figure out what is the best way to make cascading dashboard filters work for joined fields in addition to the main table fields. If possible we should implement the fix, but at minimum scope it out to understand what is needed. 

### What problem does this solve?

There is a gap in how cascading dashboard filters work right now. They currently only work within the main table in an explore, but ideally they should also work for fields from joined tables. 

There is an [internal customer thread with example video](https://lightdash.slack.com/archives/C071TU6FZ3P/p1722313055621759?thread_ts=1722211389.421979&cid=C071TU6FZ3P).

The previous work that is relevant and the decision to only filter on main table fields is here: https://github.com/lightdash/lightdash/issues/3651#issuecomment-1542239510

**Comment by ZeRego** [comment]
Example explores:

A joins B
B 
C joins D and B
D joins C and B
E

Terminology:
Scenarios are focused on the "open" filter where we run a query to fetch the possible values. "existing" filters are filters already defined in the dashboard that we want to incorporate in the query for the "open" filter.

Possible implementation in priority order:
1. if they are all from the same model, use model
2. if any existing filter models are joined to the open filter model, use the open filter model.
3. if any existing filter model joins all the other existing filter models and the open filter model, use that existing filter model
  - what happens if there are multiple matches?
4. if any existing filter model joins the open filter model, use that existing filter model and exclude the other existing models
  - what happens if there are multiple matches?

Scenarios: 
| Existing filters models | Open filter model | Is supported | Expected join logic from explore |
| --- | --- | --- | --- |
| A | A | ✅  | A (1) |
| A, A | A | ✅  | A (1) |
| B | A | 🔴 | A (2) |
| B, B | A | 🔴 | A (2) |
| C | B | 🔴 | C? (3) |
| C, D | B | 🔴 | ? (3) |
| E, A | B | 🔴 | A? + exclude E (4) |
| C, A | B | 🔴 | ? (4) |


@jpetey75 @TuringLovesDeathMetal if we implemented just 1 and 2, would it be enough? if we need to implement 3 and 4, how do we handle multiple matches?

**Comment by jpetey75** [comment]
Sorry @ZeRego I don't think I understand open vs. existing here. 

My priority is solving the problem displayed in the video linked in the description, since that would fix the customer issue and also seems like the most likely scenario others would run into. I'm not sure which direction (A joins B, B joins A) but the intended behavior would be:

1. Assuming a dashboard is built using the same model A that has a join to model B.
2. Let's say we have one permanent filter on the dashboard per model (Filter A and Filter B).
3. If I select a value for Filter A, I want to limit options in Filter B to only options that return AFTER Filter A is applied.
4. If I select a value for Filter B, I want to limit options in Filter A to only options that return AFTER Filter Bn is applied.

**Comment by ZeRego** [comment]
>  I don't think I understand open vs. existing here.

In this image, the existing filters are the first 2 and the open is the third, that is open. 

<img width="944" alt="Screenshot 2024-09-03 at 15 37 42" src="https://github.com/user-attachments/assets/fe252346-8f37-4cea-91c9-538a600afeb4">

Following you example
> If I select a value for Filter A, I want to limit options in Filter B to only options that return AFTER Filter A is applied.

Filter A = existing filter 
Filter B = open filter

> My priority is solving the problem displayed in the video linked in the description, since that would fix the customer issue and also seems like the most likely scenario others would run into. I'm not sure which direction (A joins B, B joins A) but the intended behavior would be

To solve that problem we need to understand how the models are joined. My comment is trying to cover all the scenarios we can encounter and what logic we can apply (or not).

> Assuming a dashboard is built using the same model A that has a join to model B.

Related to the previous answer, dashboards may have charts from multiple explores and filters from multiple explores.

**Comment by ZeRego** [comment]
(after a meeting with Jake)
We should implement rule 1 & 2 which have clear solutions. 
We would have to document how this works and that the joins need to be defined in the YML, even if the explore is not available as a starting point in Lightdash.
tldr: The join logic starts from the model for the filter we are editing.

**Comment by mryoho** [comment]
Ooh exciting!
